File: app/schemas/system.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel

from app.schemas.account import ORMModel


class AccountSnapshotPoint(BaseModel):
    snapshot_date: date
    balance_value: Decimal


class DailySnapshotResponse(ORMModel):
    id: int
    snapshot_date: date
    total_assets: Decimal
    total_liabilities: Decimal
    net_worth: Decimal
    investment_total: Decimal
    cash_total: Decimal


class MarketQuoteResponse(BaseModel):
    symbol: str
    name: str | None
    price: Decimal
    currency: str
    updated_at: datetime | None


class MarketSearchItem(BaseModel):
    symbol: str
    name: str


class MarketProviderStatus(BaseModel):
    provider: str
    configured: bool
    connected: bool
    message: str | None = None


class TossCredentialsUpdate(BaseModel):
    client_id: str
    client_secret: str


class HealthResponse(BaseModel):
    status: str
    db: str
    version: str


class InitializeResponse(BaseModel):
    account_types: int
    payment_methods: int
    categories: int
    message: str


class RestoreResponse(BaseModel):
    message: str
    restored_at: datetime

File: src/lib/validations/system.ts
import { z } from "zod";

import { dateSchema, decimalSchema } from "@/lib/validations/account";

export const accountSnapshotPointSchema = z.object({
  snapshot_date: dateSchema,
  balance_value: decimalSchema,
});

export const dailySnapshotResponseSchema = z.object({
  id: z.number().int(),
  snapshot_date: dateSchema,
  total_assets: decimalSchema,
  total_liabilities: decimalSchema,
  net_worth: decimalSchema,
  investment_total: decimalSchema,
  cash_total: decimalSchema,
});

export const marketQuoteResponseSchema = z.object({
  symbol: z.string(),
  name: z.string().nullable(),
  price: decimalSchema,
  currency: z.string(),
  updated_at: z.coerce.date().nullable(),
});

export const marketSearchItemSchema = z.object({
  symbol: z.string(),
  name: z.string(),
});

export const marketProviderStatusSchema = z.object({
  provider: z.string(),
  configured: z.boolean(),
  connected: z.boolean(),
  message: z.string().nullable().optional(),
});

export const tossCredentialsUpdateSchema = z.object({
  client_id: z.string(),
  client_secret: z.string(),
});

export const healthResponseSchema = z.object({
  status: z.string(),
  db: z.string(),
  version: z.string(),
});

export const initializeResponseSchema = z.object({
  account_types: z.number().int(),
  payment_methods: z.number().int(),
  categories: z.number().int(),
  message: z.string(),
});

export const restoreResponseSchema = z.object({
  message: z.string(),
  restored_at: z.coerce.date(),
});

export type MarketQuoteResponse = z.infer<typeof marketQuoteResponseSchema>;
export type MarketSearchItem = z.infer<typeof marketSearchItemSchema>;
